### PR TITLE
[Workflow] Support "retry_exceptions" of Ray tasks

### PR DESCRIPTION
**Pull request by suquark**
## Why are these changes needed?

This PR adds support for `retry_exceptions` in Ray tasks. Further integration would be add after merging https://github.com/ray-project/ray/pull/26449.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/workflow/workflow_state_from_dag.py [hunk 1]
```diff
@@ -115,17 +105,18 @@ def _node_visitor(node: Any) -> Any:
...
                     catch_exceptions = False
 
+            # We do not need to check the validness of bound options, because
+            # Ray options already check them for us.
             max_retries = bound_options.get("max_retries", 3)
-            if not isinstance(max_retries, int) or max_retries < -1:
-                raise ValueError(
-                    "'max_retries' only accepts 0, -1 or a positive integer."
-                )
+            # 'retry_exceptions' is True by default in Ray Workflow, but it is
```
Comment (iycheng) [inline python/ray/workflow/workflow_state_from_dag.py:111]: Any reason for this? SHouldn't it be aligned with normal ray?

Comment (iycheng) [reply python/ray/workflow/workflow_state_from_dag.py:111]: Also in ray, retry_exceptions can be a list of Exceptions.

Comment (suquark) [reply python/ray/workflow/workflow_state_from_dag.py:111]: @iycheng The PR for list of exceptions is not ready: https://github.com/ray-project/ray/pull/26449.

Would you think that by default `max_retries` does not work for exceptions in workflow (due to `retry_exceptions=False` by default in Ray)? I am ok with both semantics.

File: doc/source/workflows/basics.rst [hunk 2]
```diff
@@ -244,15 +244,21 @@ The following error handling flags can be either set in the task decorator or vi
     workflow.run(handle_errors.bind(r2))
 
 
-- If ``max_retries`` is given, the task will be retried for the given number of times if an exception is raised. It will only retry for the application level error. For system errors, it's controlled by ray. By default, ``max_retries`` is set to be 3.
+- Ray Workflow also supports ``max_retries`` as in Ray tasks. If ``max_retries`` is given, the task will be retried for the given number of times if an exception is raised. It will only retry for the application level error. For system errors, it's controlled by ray. By default, ``max_retries`` is set to be 3.
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:247]: I thought we disable ray layer max_retries right?

File: doc/source/workflows/basics.rst [hunk 3]
```diff
@@ -244,15 +244,21 @@ The following error handling flags can be either set in the task decorator or vi
...
 - If ``catch_exceptions`` is True, the return value of the function will be converted to ``Tuple[Optional[T], Optional[Exception]]``. This can be combined with ``max_retries`` to try a given number of times before returning the result tuple.
+- Ray Workflow also supports ``retry_exceptions`` and ``max_retries`` as in Ray tasks. If ``retry_exceptions`` is True, then the workflow task retries both task crashes and application-level errors;
```
Comment (iycheng) [inline doc/source/workflows/basics.rst:249]:
```suggestion
- Ray Workflow supports ``retry_exceptions`` as in Ray tasks. If ``retry_exceptions`` is True, then the workflow task retries both task crashes and application-level errors;
```

We've mentioned max_retries above.

Comment (iycheng) [reply doc/source/workflows/basics.rst:249]: Btw, also mention the default behavior.

File: python/ray/workflow/workflow_state_from_dag.py (hunk 1 above)
Comment (iycheng) [reply python/ray/workflow/workflow_state_from_dag.py:111]: I feel a little bit confused. If max_retries=3 and retry_exceptions=False, would we retry anything?

I think in ray, it'll retry RaySystem errors.

File: python/ray/workflow/workflow_executor.py [hunk 4]
```diff
@@ -266,50 +308,22 @@ async def _handle_ready_task(
...
-            # ---------------------- retry the task ----------------------
-            state.task_retries[task_id] += 1
-            total_retries = state.tasks[task_id].options.max_retries
-            if state.task_retries[task_id] <= total_retries:
-                logger.info(
-                    f"Retry [{workflow_id}@{task_id}] "
-                    f"({state.task_retries[task_id]}/{total_retries})"
-                )
-                state.construct_scheduling_plan(task_id)
+            if self._retry_failed_task(workflow_id, task_id, e):
```
Comment (iycheng) [inline python/ray/workflow/workflow_executor.py:311]: This is really confusing. `_retry_failed_task` actually do two things (check and retry). Maybe call it:

```
self._try_to_retry_failed_task(...)
```

File: python/ray/workflow/workflow_executor.py [hunk 5]
```diff
@@ -266,50 +308,22 @@ async def _handle_ready_task(
...
-                # NOTE: We must update the workflow status before broadcasting
-                # the exception. Otherwise, the workflow status would still be
-                # 'RUNNING' if check the status immediately after the exception.
-                wf_store.update_workflow_status(WorkflowStatus.FAILED)
-                logger.error(f"Workflow '{workflow_id}' failed due to {e}")
-                err = WorkflowExecutionError(workflow_id)
-                err.__cause__ = e  # chain exceptions
-                self._broadcast_exception(err)
-                raise err
+            if await self._catch_failed_task(workflow_id, task_id, e):
```
Comment (iycheng) [inline python/ray/workflow/workflow_executor.py:314]: ditto.

File: python/ray/workflow/workflow_executor.py (hunk 4 above)
Comment (iycheng) [reply python/ray/workflow/workflow_executor.py:311]: Also put comments about this and the next ones.

This is a big block checking different cases, so it's good that the reviewer can just read the code and don't need to jump to other functions to understand what's doing here.

**Review by fishbone — COMMENTED** [review]
Overall LG. Some questions about the semantics.

File: python/ray/workflow/workflow_state_from_dag.py (hunk 1 above)
Comment (suquark) [reply python/ray/workflow/workflow_state_from_dag.py:111]: I'll just align with Ray semantics then anyway

File: doc/source/workflows/basics.rst (hunk 2 above)
Comment (suquark) [reply doc/source/workflows/basics.rst:247]: yes, but both the API and semantics are same as Ray. for the implementation, we disabled it; but for the behavior, it is like we are doing the exactly same stuff

File: python/ray/workflow/workflow_executor.py (hunk 4 above)
Comment (suquark) [reply python/ray/workflow/workflow_executor.py:311]: ah, I think breaking the giant function would be more clear. let me bring them back again

Comment (iycheng) [reply python/ray/workflow/workflow_executor.py:311]: @suquark  I think breaking it is ok. Just the naming is too confuse. Just rename to something better reflect the purpose is good.

Comment (iycheng) [reply python/ray/workflow/workflow_executor.py:311]: > This is a big block checking different cases, so it's good that the reviewer can just read the code and don't need to jump to other functions to understand what's doing here.

What I mean is that, when read the code (even it's a function), it's very clear about what we are doing there, then that's good enough.

Btw, we should keep the function small and scope it only doing one thing.

**Comment by suquark** [comment]
I also reorganized the doc to make it more clear.

**Review by fishbone — APPROVED** [review]

**Comment by suquark** [comment]
The CI errors seem not related. I'll merge it.

**Pull request merged**